### PR TITLE
Mar 25 implement jwt auth

### PR DESCRIPTION
**Pull request by vosmerkin**
https://youtu.be/oKupiYOcE4A

https://oghakirichenko09.atlassian.net/browse/MAR-25

### CODE REVIEW

File: src/main/java/ua/tc/marketplace/jwtAuth/JwtAuthorizationFilter.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+import java.util.HashMap;
+import java.util.Map;
+
+@Component
+@AllArgsConstructor
+public class JwtAuthorizationFilter extends OncePerRequestFilter {
+    private final JwtUtil jwtUtil;
+    private final ObjectMapper mapper;
+
+//    public JwtAuthorizationFilter(JwtUtil jwtUtil, ObjectMapper mapper) {
```
Comment (OleksandrProkopenkoDev) [inline src/main/java/ua/tc/marketplace/jwtAuth/JwtAuthorizationFilter.java:29]: remove unused comments

File: src/main/java/ua/tc/marketplace/jwtAuth/JwtAuthorizationFilter.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+//        this.jwtUtil = jwtUtil;
+//        this.mapper = mapper;
+//    }
+    @Override
+    protected void doFilterInternal(HttpServletRequest request,
+                                    HttpServletResponse response,
```
Comment (OleksandrProkopenkoDev) [inline src/main/java/ua/tc/marketplace/jwtAuth/JwtAuthorizationFilter.java:35]: add `@Nonnull` on method arguments

File: src/main/java/ua/tc/marketplace/jwtAuth/JwtAuthorizationFilter.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+            if (accessToken == null ) {
+                filterChain.doFilter(request, response);
+                return;
+            }
+            System.out.println("token : "+accessToken);
+            Claims claims = jwtUtil.resolveClaims(request);
+
+            if(claims != null & jwtUtil.validateClaims(claims)){
+                String email = claims.getSubject();
+                System.out.println("email : "+email);
```
Comment (OleksandrProkopenkoDev) [inline src/main/java/ua/tc/marketplace/jwtAuth/JwtAuthorizationFilter.java:50]: remove unnecessary logs

File: src/main/java/ua/tc/marketplace/jwtAuth/JwtAuthorizationFilter.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+                Authentication authentication =
+                        new UsernamePasswordAuthenticationToken(email,"",new ArrayList<>());
+                SecurityContextHolder.getContext().setAuthentication(authentication);
+            }
+
+        }catch (Exception e){
+            errorDetails.put("message", "Authentication Error");
```
Comment (OleksandrProkopenkoDev) [inline src/main/java/ua/tc/marketplace/jwtAuth/JwtAuthorizationFilter.java:57]: Here should be constructed Application standard error message

Comment (OleksandrProkopenkoDev) [reply src/main/java/ua/tc/marketplace/jwtAuth/JwtAuthorizationFilter.java:57]: `ErrorResponse` class

File: src/main/java/ua/tc/marketplace/jwtAuth/JwtUtil.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+    public boolean validateClaims(Claims claims) throws AuthenticationException {
+        try {
+            return claims.getExpiration().after(new Date());
+        } catch (Exception e) {
+            throw e;
+        }
+    }
+
+    public String getEmail(Claims claims) {
```
Comment (OleksandrProkopenkoDev) [inline src/main/java/ua/tc/marketplace/jwtAuth/JwtUtil.java:72]: if these methods are not used, just delete them

File: src/main/java/ua/tc/marketplace/jwtAuth/JwtUtil.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+            return bearerToken.substring(jwtConfig.getTokenPrefix().length());
+        }
+        return null;
+    }
+
+    public boolean validateClaims(Claims claims) throws AuthenticationException {
+        try {
+            return claims.getExpiration().after(new Date());
+        } catch (Exception e) {
+            throw e;
```
Comment (OleksandrProkopenkoDev) [inline src/main/java/ua/tc/marketplace/jwtAuth/JwtUtil.java:68]: it is incorrect to catch e and throw e. Try to understand what happens here and throw appropriate custom exception if needed

File: src/main/java/ua/tc/marketplace/jwtAuth/JwtUtil.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+                + TimeUnit.MINUTES.toMillis(jwtConfig.getTokenExpirationAfterDays()*24*60));
+        return Jwts.builder()
+                .setClaims(claims)
+                .setExpiration(tokenValidity)
+                .signWith(SignatureAlgorithm.HS256, jwtConfig.getSecretKey())
+                .compact();
+    }
+
+    private Claims parseJwtClaims(String token) {
+        return Jwts.parser()
```
Comment (OleksandrProkopenkoDev) [inline src/main/java/ua/tc/marketplace/jwtAuth/JwtUtil.java:34]: `parser` is deprecated. Try to refactor and use actual methods

File: src/main/java/ua/tc/marketplace/jwtAuth/JwtUtil.java [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+                .setSigningKey(jwtConfig.getSecretKey())
```
Comment (OleksandrProkopenkoDev) [inline src/main/java/ua/tc/marketplace/jwtAuth/JwtUtil.java:35]: `setSigningKey` is deprecated, try to replace by new approach

File: src/main/java/ua/tc/marketplace/jwtAuth/JwtUtil.java [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+        Claims claims = Jwts.claims().setSubject(user.getEmail());
+        claims.put("firstName",user.getFirstName());
+        claims.put("lastName",user.getLastName());
+        Date tokenCreateTime = new Date();
+        Date tokenValidity = new Date(tokenCreateTime.getTime()
+                + TimeUnit.MINUTES.toMillis(jwtConfig.getTokenExpirationAfterDays()*24*60));
+        return Jwts.builder()
+                .setClaims(claims)
+                .setExpiration(tokenValidity)
+                .signWith(SignatureAlgorithm.HS256, jwtConfig.getSecretKey())
```
Comment (OleksandrProkopenkoDev) [inline src/main/java/ua/tc/marketplace/jwtAuth/JwtUtil.java:29]: `signWith` is deprecated, try to replace with something new

File: src/main/java/ua/tc/marketplace/jwtAuth/JwtUtil.java [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+public class JwtUtil {
+    private final JwtConfig jwtConfig;
+
+    public String createToken(User user) {
+        Claims claims = Jwts.claims().setSubject(user.getEmail());
+        claims.put("firstName",user.getFirstName());
+        claims.put("lastName",user.getLastName());
+        Date tokenCreateTime = new Date();
+        Date tokenValidity = new Date(tokenCreateTime.getTime()
+                + TimeUnit.MINUTES.toMillis(jwtConfig.getTokenExpirationAfterDays()*24*60));
```
Comment (OleksandrProkopenkoDev) [inline src/main/java/ua/tc/marketplace/jwtAuth/JwtUtil.java:25]: make that numbers (24*60) as CONSTANT

File: src/main/java/ua/tc/marketplace/jwtAuth/JwtUtil.java [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+
+    public Claims resolveClaims(HttpServletRequest req) {
+        try {
+            String token = resolveToken(req);
+            if (token != null) {
+                return parseJwtClaims(token);
+            }
+            return null;
+        } catch (ExpiredJwtException ex) {
+            req.setAttribute("expired", ex.getMessage());
```
Comment (OleksandrProkopenkoDev) [inline src/main/java/ua/tc/marketplace/jwtAuth/JwtUtil.java:47]: need to understand how is this working. And make it work as we need. We need to send to Client correct error message of our standard form. And print into app's console only one line with `ERROR` tag and actual error message

File: src/main/java/ua/tc/marketplace/jwtAuth/JwtUtil.java (hunk 6 above)
Comment (vosmerkin) [reply src/main/java/ua/tc/marketplace/jwtAuth/JwtUtil.java:69]: moved exception to calling method for now. But I still dont like it, will again search for better way to catch expiration exception

**Review by OleksandrProkopenkoDev — APPROVED** [review]

**Review by NikolayJRStudent — APPROVED** [review]

**Pull request merged**